rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+- **Feature Name:** Loop-contract synthesis
+- **Feature Request Issue:** <https://github.com/model-checking/kani/issues/2214>
+- **RFC PR:** <https://github.com/model-checking/kani/pull/2215>
+- **Status:** Under Review
+- **Version:** 0
+- **Proof-of-concept:** <https://github.com/qinheping/kani/tree/kani-synthesizer>
+
+## Summary
+
+An new option that allows users to verify programs without unwinding loops by synthesizing loop contracts for those loops.
+
+
+## User Impact
+Currently Kani does not support proving correctness (i.e. assertions never fail) on programs with unbounded control flow (e.g. loops with dynamic bounds). Kani unrolls all unbounded loops until a few iterations and then verifies this unrolled program — it thus provides a much weaker guarantee on correctness.
+
+The loop-contract synthesizer will synthesize loop contracts for loops without unwinding numbers specified. With the synthesized loop contracts, CBMC abstracts those loops as non-loop structures that can cover arbitrary iterations of the loops. As a result, with the synthesizer proposed in this RFC, Kani can verify programs without unwind loops, and provide a stronger guarantee on correctness. For example, the number of executed iterations of the loop in the following harness is dynamically bounded by an unbounded variable `y`; hence no unwinding number can guarantee the correctness of the loop.
```

Comment:
In the text you mention "correctness of the loop" but in the examples, there is nothing inside the loop that gets proven. What happens if there is an assertion inside the loop or an array access? Do we still prove their correctness? If so, can you please add such an example?

Reply:
It should be "Only the unwinding number `i32::MAX` can guarantee to cover all iteration of the loop, and hence satisfies the unwinding assertions.".